text/0000-time.md

```diff
@@ -0,0 +1,359 @@
+
+# Summary
+[summary]: #summary
+
+This document introduces a time sub-system into Twizzler. The time sub-system lays down the foundation for Twizzler's notion of time and interfaces for user space programs. We introduce a trait that abstracts the hardware, new system calls in the kernel, and types to represent clocks.
+
+# Motivation
+[motivation]: #motivation
+
+We care about supporting user space programs, particularly Rust's standard library. A big part of that is providing APIs for time. User programs require services for time to do useful things such as benchmarking.
+
+Rust exposes an interface for time to user space through various types.
+
+* [Duration](https://doc.rust-lang.org/std/time/struct.Duration.html): represents span of time (sec and ns, like timespec)
+* [Instant](https://doc.rust-lang.org/std/time/struct.Instant.html): a monotonic non-decreasing type that is implemented using [os-specific system calls](https://doc.rust-lang.org/std/time/struct.Instant.html#underlying-system-calls)
+    * e.g. `clock_gettime(CLOCK_MONOTONIC)` for Linux
+* [SystemTime](https://doc.rust-lang.org/std/time/struct.SystemTime.html): represents an anchor in time from the unix epoch, non-monotonic
+    * e.g. `clock_gettime(CLOCK_REALTIME)` for Linux
+
+To support the current APIs in Rust, we need to provide an implementation of a monotonic clock and system clock. Twizzler currently provides [stubs](https://twizzler-operating-system.github.io/nightly/doc/twizzler_abi/time/index.html) so that Rust can call into the kernel. 
+
+The outcome of this work is to: 
+ * Provide strong support for user space Rust applications using its time APIs
+* Develop interfaces for Twizzler native programs
+* Develop standard interfaces for the kernel to manage the hardware used for clocks.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The time interface exposed to users is traditional in the sense that all mechanisms in the kernel hide behind system calls. However, user programs are free to use libraries built on top of system calls, as is done so in the Rust standard library.
+
+One could imagine a simple Rust program that reads the elapsed time from a monotonic clock, ported over to Twizzler as so:
+```rust
+use std::time::Instant;
+fn main() {
+    // reads a moment in time
+    let now = Instant::now();
+    
+    // elapsed time since now
+    // this returns a Duration type
+    let elapsed_time = now.elapsed();
+    println!("Running main() took {} nanoseconds.", elapsed_time.as_nanos());
+}
+```
+
+## Clock and ClockInfo
+
+To support the above functionality, Twizzler exposes a `Clock` abstraction to user space, and APIs revolve around this and `ClockInfo`. A `Clock` is a logical abstraction exposed to users that serves a particular purpose and ticks at a particular pace.
+
+```rust
+struct Clock {
+    info: ClockInfo,
+    usage: ClockUsage,
+    id: u64 // leads us to the clock source
```

Comment:
I'd like this to be a type, something like
`#[repr(transparent)] struct ClockId(u64)`